### ISSUE TITLE
How do the time correction servers work?

### ISSUE THREAD

**Issue by jaybeers** [KeeTrayTOTP/KeeTrayTOTP#31]
This is really just a documentation question, but...can you explain how the "time correction" servers work?  I don't mean how to set them up (that's covered in the documentation), I mean what do they actually _do_?  It's not NTP; based on https://sourceforge.net/p/traytotp-kp2/tickets/7 what it's looking for is an HTTPS URI as input, but what is it looking for?

At first I thought it was retrieving the `Date` response header, but I tried two different URIs that both return that header and one worked, one didn't.

And by the way, yes I know that this is really a question for "upstream" Tray TOTP, but that SourceForge project seems dead and I was kinda hoping you might be able to make more sense of the source code than I could (I've never done anything with C# before).  Thanks in advance if you can shed any light on this...or if not, thanks for reading this far at least.  :)

**Comment by jaybeers** [comment]
Actually, I just found https://github.com/victor-rds/KeeTrayTOTP/blob/master/Libraries/TimeCorrectionProvider.cs#L118 which leads me to believe that I was right after all (it _is_ retrieving the `Date` header from the HTTP response); that's odd though, because then that doesn't explain why it doesn't work with some of the URIs I tried, which do return that header (and the date is in the same format).  For instance, `https://www.google.com/` works, but `https://github.com/` doesn't.

**Comment by robinvanpoppel** [comment]
Hi John,

I am not familiar with the time correction stuff. I also looked at the headers returned by `https://www.google.com/` and `https://github.com/`. I at first typed `https://www.github.com`, which gave me a redirect with the following 3 headers (and missing the Date header):
```
HTTP/1.1 301 Moved Permanently
Content-length: 0
Location: https://github.com/
```
Do you see any difference in behavior when using `https://www.github.com` vs `https://github.com`?

**Comment by jaybeers** [comment]
No, I don't (surprisingly); I tried all four of these (with and without the `www.` and `/`) and none of them works in TrayOTP (although they do have different header results when retrieved manually, as you'd expect).  For the record, I'm using `curl -is <URI> | head -n 30` for the manual information gathering (I don't suspect that matters, but I guess you never know).

```
https://www.github.com
https://www.github.com/
https://github.com
https://github.com/
```

**Comment by robinvanpoppel** [comment]
I've tried verification at my home network with the four url's in your comment and they seem to work. Assuming you are running the latest version if this plugin, I've got two additional questions.
* Are you running behind a proxy? 
* Are you using this plugin on windows or perhaps via mono? It might be that the WebClient in mono is behaving differently.

**Comment by jaybeers** [comment]
I'm not behind a proxy per se; I'm behind NAT but it seems like I wouldn't be the only one with the problem if it were that simple.  I'm using version 2.0.0.5 of the upstream plugin, which is the newest available version.  However, I _am_ using the plugin via Mono.

On a side note: should I consider switching over to using this fork instead of the original/upstream version?  When I was originally looking stuff up I just went with upstream because that's just what you do, and I'm not interested in Steam MFA.  But it looks like the upstream codebase is all but abandoned and this one is still being looked after.  :)  Just a thought; if I were to drop in the newest .plgx file from this repo, would it be a drop-in replacement (i.e. would I have to clear out my current TOTP setups and regenerate them)?

**Comment by victor-rds** [comment]
> Just a thought; if I were to drop in the newest .plgx file from this repo, would it be a drop-in replacement (i.e. would I have to clear out my current TOTP setups and regenerate them)?

AFAIK it works, since it uses the same custom fields , like `TOTP Settings` when I first made this fork that was my case, everything worked for me

**Comment by robinvanpoppel** [comment]
@jaybeers let me know if you run into any problems with this plugin as a drop-in replace for the non-maintaned one. What would you like to do with this issue?

**Comment by jaybeers** [comment]
Well, unfortunately I'm not able to test because the KeePass versions I'm using (2.36 and 2.38 on two separate machines...three cheers for LTS releases!) are too old for this plugin.  :(  I don't want you guys to spend a bunch of energy on this on my account, when based on what you said it sounds like it works for you anyway.  If I upgrade to a newer version of KeePass and I can use your fork of this plugin and the problem still persists, then it might be investigating as being related to the Mono implementation I suppose.  Thanks for looking into it!  If somebody else gets interested in this issue, feel free to let me know if you want me to test something with Mono.  :)

**Comment by jaybeers** [comment]
So, it's only been a couple of years but I finally got in a position where I could move to the new LTS release of my OS and this problem is resolved for me (KeePass 2.44 / KeeTrayTOTP 0.108).  Sorry for keeping this on your plate for so long, but as far as I'm concerned this issue can be closed, now.  Thanks!
